model-engine/model_engine_server/inference/post_inference_hooks.py

```diff
@@ -76,6 +83,69 @@ def handle(
                 assert 200 <= res.status_code < 300
 
 
+class LoggingHook(PostInferenceHook):
+    def __init__(
+        self,
+        endpoint_name: str,
+        bundle_name: str,
+        user_id: str,
+        endpoint_id: Optional[str],
+        endpoint_type: Optional[ModelEndpointType],
+        bundle_id: Optional[str],
+        labels: Optional[Dict[str, str]],
+    ):
+        super().__init__(endpoint_name, bundle_name, user_id)
+        self._endpoint_id = endpoint_id
+        self._endpoint_type = endpoint_type
+        self._bundle_id = bundle_id
+        self._labels = labels
+
+    def handle(
+        self,
+        request_payload: EndpointPredictV1Request,
+        response: Dict[str, Any],
+        task_id: Optional[str],
+    ):
+        sts_client = boto3.client("sts", region_name=infra_config().default_region)
+        assumed_role_object = sts_client.assume_role(
+            RoleArn=infra_config().firehose_role_arn,
+            RoleSessionName="AssumeRoleSession1",
+        )
+        credentials = assumed_role_object["Credentials"]
+        session = boto3.Session(
+            aws_access_key_id=credentials["AccessKeyId"],
+            aws_secret_access_key=credentials["SecretAccessKey"],
+            aws_session_token=credentials["SessionToken"],
+        )
+        firehose_client = session.client("firehose", region_name=infra_config().default_region)
```

Comment:
okay, I think I got my wires crossed a bit earlier. Rather than requiring the service to assume a new IAM role, we should expand the existing role's perms to enable publishing to this stream. 

Effectively, we want to add the permissions declared [here](https://github.com/scaleapi/Terracode-ML/pull/596/files#diff-f2db50a56315df230a5a73d6935fa9f8f4700f61fe4ac724c034c491136333e1R655-R660) to the `ml-worker` role already assumed by endpoints. 

We can then follow the existing pattern for instantiating AWS clients ([s3 example](https://github.com/scaleapi/llm-engine/blob/14635b0be69c5331b90f2a50f263b57e9558a714/model-engine/model_engine_server/infra/gateways/s3_filesystem_gateway.py#L15)) inside the implementations.

Reply:
I'm sorry I caught this later @tiffzhao5. We might need to redo some of the TF prs. We also probably don't need to introduce a new role ARN in this PR: https://github.com/scaleapi/models/pull/8808

Reply:
By adding the permissions to the `ml-worker` role already assumed by the endpoints, are you referring to this? https://github.com/scaleapi/Terracode-Infra/pull/2357/files because I've tried this already and it didn't work. This was because the firehose stream I created lives in the ml-account, while ml-worker lives in the scale-prod account, so the solution was to create a new IAM role in the ml-account (and serving) and have ml-worker assume that role.

Did I understand your proposal correctly? Also, I'm open to other solutions -- doing it this way was advised by platform so I went with it given little context on this.

Reply:
I think your original proposal and explanation is correct. Looks like Kinesis Firehose doesn't support `resource-based` policies ([link](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_aws-services-that-work-with-iam.html)). If it did, we would then just attach a policy to the firehose to allow cross-account `ml-worker` to publish to the stream.

Looks like your approach is required. I would just leave a comment in the code explaining why an additional role is required. 

In summary, our current setup is "service deployed in ML account assumes role from Prod account to assume a role in ML account to publish to ML account resources" 😿. Not a problem to fix in this PR.

Reply:
Another option is to instead move the streams into the `Prod` account, but that probably isn't worth it just for IAM reasons.